src/declarativewebcontainer.cpp

```diff
@@ -59,14 +60,26 @@ DeclarativeWebContainer::DeclarativeWebContainer(QQuickItem *parent)
     , m_completed(false)
     , m_initialized(false)
 {
-    m_webPages.reset(new WebPages(this));
     setFlag(QQuickItem::ItemHasContents, true);
+    setPrivateMode(m_settingManager->autostartPrivateBrowsing());
+
+    m_normalWebPages.reset(new WebPages(this));
+    m_privateWebPages.reset(new WebPages(this));
+
+    m_persistentTabModel.reset(new PersistentTabModel());
+    m_privateTabModel.reset(new PrivateTabModel());
+
+    setTabModel(privateMode() ? m_privateTabModel.data() : m_persistentTabModel.data());
+
+    setWebPages();
+
     connect(DownloadManager::instance(), SIGNAL(initializedChanged()), this, SLOT(initialize()));
     connect(DownloadManager::instance(), SIGNAL(downloadStarted()), this, SLOT(onDownloadStarted()));
     connect(QMozContext::GetInstance(), SIGNAL(onInitialized()), this, SLOT(initialize()));
     connect(this, SIGNAL(portraitChanged()), this, SLOT(resetHeight()));
     connect(this, SIGNAL(enabledChanged()), this, SLOT(handleEnabledChanged()));
 
+    connect(this, SIGNAL(privateModeChanged()), this, SLOT(updateMode()));
```

Comment:
updateMode() should be called directly from setPrivateMode before emitting change signal. When change signal is emitted all properties should match to that mode.